src/platform_impl/web/web_sys/mod.rs

```diff
@@ -60,35 +60,83 @@ impl WindowExtWebSys for Window {
     }
 }
 
-pub fn window_size() -> LogicalSize<f64> {
-    let window = web_sys::window().expect("Failed to obtain window");
-    let width = window
-        .inner_width()
-        .expect("Failed to get width")
-        .as_f64()
-        .expect("Failed to get width as f64");
-    let height = window
-        .inner_height()
-        .expect("Failed to get height")
-        .as_f64()
-        .expect("Failed to get height as f64");
-
-    LogicalSize { width, height }
-}
-
 pub fn scale_factor() -> f64 {
     let window = web_sys::window().expect("Failed to obtain window");
     window.device_pixel_ratio()
 }
 
-pub fn set_canvas_size(raw: &HtmlCanvasElement, size: Size) {
+/// Gets the size of the content box of `element` based on CSS.
+///
+/// Returns `None` if the element isn't in the DOM.
+pub fn inner_size(raw: &HtmlCanvasElement) -> Option<LogicalSize<f64>> {
+    let window = web_sys::window().unwrap();
+    let document = window.document().unwrap();
+    if !document.contains(Some(raw)) {
+        return None;
+    }
+
+    // Use `getBoundingClientRect` instead of the width and height properties because it doesn't round to the nearest integer.
+    let rect = raw.get_bounding_client_rect();
+    let style = window
+        .get_computed_style(raw)
+        .unwrap()
+        .expect("`getComputedStyle` returned `None`");
+
+    let prop = |name| -> f64 {
+        let value = style.get_property_value(name).unwrap();
```

Comment:
Might be missing something but what guarantees that the property values are going to be in px?

Reply:
I didn't think about this at the time, but after doing some digging it's *almost* guaranteed - the properties will always be in `px` unless the canvas' `display` is `none` or `contents`.

<details>
<summary>Spec-digging:</summary>

First, [getComputedStyle](https://drafts.csswg.org/cssom/#dom-window-getcomputedstyle) defines the values of the properties it returns as such:

> with the value being the [resolved value](https://drafts.csswg.org/cssom/#resolved-value) computed for obj using the style rules associated with doc.

For the padding properties, the resolved value is defined as:

> If the property applies to the element or pseudo-element and the [resolved value](https://drafts.csswg.org/cssom/#resolved-value) of the ['display'](https://drafts.csswg.org/css-display-3/#propdef-display) property is not ['none'](https://drafts.csswg.org/css-display-3/#valdef-display-none) or ['contents'](https://drafts.csswg.org/css-display-3/#valdef-display-contents), then the resolved value is the [used value](https://drafts.csswg.org/css-cascade-5/#used-value). Otherwise the resolved value is the [computed value](https://drafts.csswg.org/css-cascade-5/#computed-value).

The used value is defined as:

> The used value is the result of taking the [computed value](https://drafts.csswg.org/css-cascade-5/#computed-value) and completing any remaining calculations to make it the absolute theoretical value used in the formatting of the document.

This isn't really a formal definition, but I think that this should mean that the used value of the padding is an [absolute length](https://drafts.csswg.org/css-values-4/#absolute-lengths).

No special case is defined for the border width properties, so they fall back to the default:

> The [resolved value](https://drafts.csswg.org/cssom/#resolved-value) is the [computed value](https://drafts.csswg.org/css-cascade-5/#computed-value).

The computed value is defined as:

> The computed value is the result of resolving the [specified value](https://drafts.csswg.org/css-cascade-5/#specified-value) as defined in the “Computed Value” line of the property definition table, generally absolutizing it in preparation for [inheritance](https://drafts.csswg.org/css-cascade-5/#inheritance).

The padding properties are defined [here](https://drafts.csswg.org/css-box/#padding-physical), and their computed values are defined as:

> a computed [\<length-percentage\>](https://drafts.csswg.org/css-values-4/#typedef-length-percentage) value

`<length-percentage>` just means `<length>` or `<percentage>`.

The border width properties are defined [here](https://drafts.csswg.org/css-backgrounds/#the-border-width), and their computed values are defined as:

> absolute length; zero if the border style is [none](https://drafts.csswg.org/css-backgrounds/#valdef-line-style-none) or [hidden](https://drafts.csswg.org/css-backgrounds/#valdef-line-style-hidden)

So, I think both of these properties' resolved values should be absolute lengths as long as the canvas' `display` isn't `none` or `contents`, since:
- the border width's computed value is defined to always be an absolute length.
- as long as the canvas' display isn't `none` or `contents`, the padding's resolved value is its used value, which is absolute.

Then, how these values are converted to strings is defined in [getPropertyValue](https://www.w3.org/TR/cssom/#dom-cssstyledeclaration-getpropertyvalue), which calls into the [serialize a CSS value](https://www.w3.org/TR/cssom/#serialize-a-css-value) to do the actual conversion. `<length>` is defined to be serialized as:

> The <number> component serialized as per <number> followed by the unit in its canonical form as defined in its respective specification.

Both the properties' resolved values were [absolute lengths](https://drafts.csswg.org/css-values-4/#absolute-lengths) when the `display` wasn't `none` or `contents`, and:

> All of the absolute length units are [compatible](https://drafts.csswg.org/css-values-4/#compatible-units), and [px](https://drafts.csswg.org/css-values-4/#px) is their [canonical unit](https://drafts.csswg.org/css-values-4/#canonical-unit).

So, I think the values should always be in `px` when the `display` isn't `none` or `contents`.

When it is, though, the `padding` can be a `<percentage>`, which is serialized as:

> The <number> component serialized as per <number> followed by the literal string "%" (U+0025).

So, this won't work when `display` is `none` or `contents`.

</details>

So, I've changed `prop` to return 0 for padding when `display` is `none` or `contents`, since I don't think it's worth the effort to compute the percentage for a hidden canvas.

Also, removing the `px` by slicing is dodgy, so I've replaced that with a `strip_suffix("px")` and added a bit more context to the panic message.

Reply:
I wasn't aware of that. Thanks! :+1: